Adding I/O for direct analysis of near-surface wind gust for RRFS-based 3DRTMA

GangZhao-NOAA opened the pull request:
<!-- PLEASE READ -->
<!--
Before opening a PR, please note these guidelines:

- Each PR should only address ONE topic and have an associated issue
- No hardcoded or paths to personal directories should be present
- No temporary or backup files should be committed
- Any code that was disabled by being commented out should be removed
-->

**Description**

<!-- Please include relevant motivation and context. -->
<!-- Please include a summary of the change and which issue is fixed. -->
<!-- List any dependencies that are required for this change. -->
To improve the analysis of the near-surface wind gust in 3DRTMA, the observations of near-surface wind gust would be analyzed directly in GSI (3DVar and Hybrid 3DEnVar), instead of being a derived product from the near-surface wind analysis.

Since the core subroutines for direct variational assimilation of wind gust (e.g., setupgust.f90, intgust.f90, stpgust.f90, etc.) had already been implemented in GSI for 2DRTMA, so in the work the development in GSI mainly focus on adding I/O of 2-D wind gust firstguess and analysis fields for RRFS-based 3DRTMA, and some minor modifications in observation and background error for wind gust, options to control the analysis of wind gust, etc.
<!-- Please provide reference to the issue this pull request is addressing. -->
<!-- For e.g. Fixes #IssueNumber -->
This PR is to address the issue #726 : Adding I/O for direct analysis of near-surface wind gust for RRFS-based 3DRTMA 
Fixes #726
**Type of change**

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

**How Has This Been Tested?**

<!-- Please describe the tests that you ran to verify your changes and on the platforms these tests were conducted. -->
<!-- Provide instructions so we can reproduce. -->
<!-- Please also list any relevant details for your test configuration -->
  
**Checklist**

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing tests pass with my changes
      tested with a real case - 2024-02-20_12:00Z, 
1. dry-run (using my updated GSI code with wind gust analysis, but actually no wind gust obs is analyzed, so-called dryrun) is compared to control-run (original GSI code running without wind gust obs): the results are identical. This indicates that if without analyzing wind-gust obs, then the updated code generates the analysis identical to the analysis of original/control code. Or say, the added code does not have influence on the other part of code.
2. real case run with updated GSI code to analyze the obs of wind gust:
The following figure shows the used observations of near-surface wind gust:
![var_obs_2024022012_gust_used_maprll_datll_reg_ncf](https://github.com/NOAA-EMC/GSI/assets/53267411/ecbe479a-03c6-490f-a179-9e0027291468)
the following figure shows the analysis increments:
![GUST_hyb_hwllp90_corptuned_inc_incrintrp_maprll_datrll_reg_grb2](https://github.com/NOAA-EMC/GSI/assets/53267411/a01fca0d-dc1f-438b-b8eb-e624de35a631)
- [x] Any dependent changes have been merged and published
- [x] Regression tests on WCOSS2 (Cactus) and Hera (Rocky-8) : my updated GSI commit [#f91f247d](https://github.com/GangZhao-NOAA/GSI/commit/f91f247dde65dea0b659c0b78d747433d9ca9559)) vs control/original GSI code (commit [#6d9ebbb7](https://github.com/NOAA-EMC/GSI/commit/6d9ebbb7896b92a93959ce63c7a1ad9e9a0aab4f))
Here is the reports of the regression tests on WCOSS2 (Cactus):
~~~
[gang.zhao@clogin02:build] (feature/windgust_in_3dvar_for_3drtma)$ ctest -j 7
Test project /lfs/h2/emc/da/save/gang.zhao/WorkDir/ProdGSI_Dev/gsi_dev/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_glbens
    Start 4: netcdf_fv3_regional
    Start 5: hafs_4denvar_glbens
    Start 6: hafs_3denvar_hybens
    Start 7: global_enkf
1/7 Test #4: netcdf_fv3_regional ..............   Passed  483.15 sec
2/7 Test #3: rrfs_3denvar_glbens ..............   Passed  486.74 sec
3/7 Test #7: global_enkf ......................   Passed  850.98 sec
4/7 Test #2: rtma .............................   Passed  970.28 sec
5/7 Test #6: hafs_3denvar_hybens ..............   Passed  1152.82 sec
6/7 Test #5: hafs_4denvar_glbens ..............   Passed  1213.93 sec
7/7 Test #1: global_4denvar ...................   Passed  1683.16 sec

100% tests passed, 0 tests failed out of 7

Total Test time (real) = 1683.19 sec
~~~
Here is the reports of the regression tests on Hera (Rocky8):
~~~
(base) [Gang.Zhao@hfe11:build] (feature/windgust_in_3dvar_for_3drtma)$ ctest -j 7
Test project /scratch1/NCEPDEV/da/Gang.Zhao/ProdGSI_dev/gsi_dev/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_glbens
    Start 4: netcdf_fv3_regional
    Start 5: hafs_4denvar_glbens
    Start 6: hafs_3denvar_hybens
    Start 7: global_enkf
1/7 Test #4: netcdf_fv3_regional ..............   Passed  491.53 sec
2/7 Test #3: rrfs_3denvar_glbens ..............***Failed  495.27 sec
3/7 Test #2: rtma .............................   Passed  982.45 sec
4/7 Test #6: hafs_3denvar_hybens ..............   Passed  1168.99 sec
5/7 Test #7: global_enkf ......................   Passed  1239.77 sec
6/7 Test #5: hafs_4denvar_glbens ..............***Failed  1347.87 sec
7/7 Test #1: global_4denvar ...................   Passed  1974.45 sec

71% tests passed, 2 tests failed out of 7

Total Test time (real) = 1974.91 sec

The following tests FAILED:
          3 - rrfs_3denvar_glbens (Failed)
          5 - hafs_4denvar_glbens (Failed)
Errors while running CTest
Output from these tests are in: /scratch1/NCEPDEV/da/Gang.Zhao/ProdGSI_dev/gsi_dev/build/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
(base) [Gang.Zhao@hfe11:build] (feature/windgust_in_3dvar_for_3drtma)$ ctest -R rrfs_3denvar_glbens
Test project /scratch1/NCEPDEV/da/Gang.Zhao/ProdGSI_dev/gsi_dev/build
    Start 3: rrfs_3denvar_glbens
1/1 Test #3: rrfs_3denvar_glbens ..............   Passed  430.52 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) = 430.55 sec
(base) [Gang.Zhao@hfe11:build] (feature/windgust_in_3dvar_for_3drtma)$ ctest -R hafs_4denvar_glbens
Test project /scratch1/NCEPDEV/da/Gang.Zhao/ProdGSI_dev/gsi_dev/build
    Start 5: hafs_4denvar_glbens
1/1 Test #5: hafs_4denvar_glbens ..............   Passed  1225.37 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) = 1225.39 sec
~~~
**Note**: 
_When I was running the regression tests, GSI code was just updated to the latest commit [#b53740a7](https://github.com/GangZhao-NOAA/GSI/commit/f91f247dde65dea0b659c0b78d747433d9ca9559). Considering the frequent update in EMC GSI code recently and saving the time, after this PR has been reviewed and approved by peer-reviewers, I will update the code to latest EMC GSI commit, then re-run the regression tests for final approval.

GangZhao-NOAA committed:
Added code for (mainly in I/O) analysis of wind gust in GSI-3DVar i
for RRFS-based 3DRTMA (fv3-lam firstguess).

GangZhao-NOAA committed:
Merge remote-tracking branch 'upstream/develop' (commit #7c4a5718)
into feature/windgust_in_3dvar_for_3drtma

GangZhao-NOAA committed:
Merge remote-tracking branch 'upstream/develop' (commit#fca6bea4e5c737efe53fd053ca9cc69033f847f7)
 into feature/windgust_in_3dvar_for_3drtma

GangZhao-NOAA committed:
Merge remote-tracking branch 'upstream/develop' (commit #dfb958f) into
 feature/windgust_in_3dvar_for_3drtma

GangZhao-NOAA committed:
Merge remote-tracking branch 'upstream/develop' (commit #2167bc93) into
 feature/windgust_in_3dvar_for_3drtma (4884b183)

GangZhao-NOAA committed:
Merge remote-tracking branch 'upstream/develop' (commit #f7e93abf)
into feature/windgust_in_3dvar_for_3drtma

GangZhao-NOAA committed:
Merge remote-tracking branch 'upstream/develop' (commit #6d9ebbb7)
 into feature/windgust_in_3dvar_for_3drtma

AnnetteGibbs-NOAA reviewed (approved):
Looks good Gang.

TingLei-NOAA commented:
@GangZhao-NOAA   Updated version:   where is  the updated surface wind stored ?  if they are in fv3_sfcdata, should this change cause regional regression tests fail if the regression tests compare those files?

TingLei-NOAA commented on src/gsi/rapidrefresh_cldsurf_mod.f90:
```diff
@@ -447,6 +458,10 @@ subroutine init_rapidrefresh_cldsurf
     corp_howv           = 0.42_r_kind                 ! 0.42 meters (default)
     hwllp_howv          = 170000.0_r_kind             ! 170,000.0 meters (170km as default for 3DRTMA, 50km is used in 2DRTMA)
     i_howv_3dda         = 0                           ! no analysis of significant wave height (howv) in 3D analysis (default)
+    corp_gust           = -3.00_r_kind                ! default to use the value (3.0 m/s) set in m_berror_stats_reg.f90
```
The comment seems not explaining why they are first initialized to negative values.

GangZhao-NOAA replied:
@TingLei-NOAA 
Hi Ting,
The negative value means to use the default value which has been hardwired in GSI code (subroutine berror_read_wgt_reg). I will add more explanation in the comment to make it more clear.
Thanks,
-Gang

TingLei-NOAA commented on src/gsi/m_berror_stats_reg.f90:
```diff
@@ -830,6 +831,10 @@ subroutine berror_read_wgt_reg(msig,mlat,corz,corp,hwll,hwllp,vz,rlsig,varq,qopt
         do i=0,mlat+1
            hwllp(i,n)=hwll(i,1,nrf3_q)
         end do
+        if ( l_rtma3d ) then  ! tuning stddev and length for analysis of gust in 3drtma test run
```
If the users forget to set values (corp_gust and/or hwllp_gust),  the code will use the above (default) q values without any warning message and the user might be unable to notice this .  Would it be better to let GSI abort with explanation if negative values are encountered here?

GangZhao-NOAA replied:
@TingLei-NOAA
Hi Ting,

The purpose for introducing corp_gust and hwllp_gust is to give the user the options to tune the background error (co)variances. The default values  have been tuned to be optimal for general cases and good enough to be used in most cases, and user does not need to concern about it. When corp_gust or hwllp_gust is negative, it means to use the default values for analysis of wind gust, so GSI does not need to abort.  

I will add more clear explanation and a printout to remind the user of the used values of the parameters of background and observations error (co)variances for wind gust in GSI code. Thank you for pointing it out!

-Gang

TingLei-NOAA replied:
@GangZhao-NOAA  But on line 832, it is the hwll of q on the first level being used as the default for wind gust, which could be tuned for considerations on q itself.  So, is such default reliable ?

GangZhao-NOAA replied:
@TingLei-NOAA  This is a good point. Using the de-correlation length scale of q for wind gust is originally used in 2D-RTMA. So it is kind of legacy code from 2DRTMA, and since 2DRTMA is still running in operation, it has to be kept there. 

So, I agree with you that using the value for q might not be a good choice for gust for different DA system. This is the reason why I introduce namelist option "hwllp_gust" into GSI so that user could tune it, instead of using the fixed value for q. For example,  for 3DRTMA, the de-correlation length of q at bottom level is reduced to very short scale (1/8 of regular value), which is too short for wind gust, so in my test run of 3DRTMA, I set hwllp_gust=90000.0 (90km) in the gsiparm.anl.  Notice that hwllp_gust is just a real-type number, not an array, so this user-specified de-correlation length scale for wind gust has to be homogenous (universal value for everywhere), could not change with, such as with latitude as in GSI static B-matrix files.

Thank you!

TingLei-NOAA commented on src/gsi/read_prepbufr.f90:
```diff
@@ -2556,6 +2557,7 @@ subroutine read_prepbufr(nread,ndata,nodata,infile,obstype,lunout,twindin,sis,&
 !                need to find out gustoe
 !                gustoe=1.8
                  gustoe=1.0
+                 if ( l_rtma3d .and. oerr_gust > 0.0_r_kind ) gustoe = oerr_gust
```
the same concern,  if i_gust_3dda=1 but oerr_gust is not set in the namelist , this line would be not executed that the user might be unable to find .   And, here, maybe i_gust_3dda is a better parameter to be tested than l_rtma3d.

GangZhao-NOAA replied:
@TingLei-NOAA 
Hi Ting,
Similar as the setup for parameters of background error (co)variances for wind gust, in most cases, esp. in 2DRTMA run, the observation error of wind gust (gustoe) is the default value which is hardwired in GSI code (1.0 m/s). Usually user does not need to set this parameter. Even if user would like to tune the obs error of wind gust, it is only for 3drtma run (definitely not in any 2DRTMA run), that's why l_rtma3d is used here to make sure obs error of wind gust is changed only if it is 3DRTMA run.

I will following your suggestion to add a print-out of the observation error of wind gust to remind the user. 

Thank you!

-Gang

TingLei-NOAA replied:
@GangZhao-NOAA  Thanks for the explanation.  That is ok if it is restricted 3drtma ( in the future, if, say, rrfs need this, it could be changed ).   And no need to add a warning message, since it is in a loop and would produce maybe too many duplicated messages.

GangZhao-NOAA replied:
@TingLei-NOAA Yes, definitely we can set up the observation of error of wind gust also for other DA systems (such as rrfsf-da) if gust is analyzed in that DA system. We can set different values of obs error of gust with respect to different DA systems.
Thank you!

TingLei-NOAA commented on src/gsi/read_prepbufr.f90:
```diff
@@ -2577,6 +2579,7 @@ subroutine read_prepbufr(nread,ndata,nodata,infile,obstype,lunout,twindin,sis,&
                    if ((kx==188).or.(kx==288) .or.(kx==195) .or.(kx==295)) then
 !                     gustoe=2.5
                       gustoe=1.0
+                      if ( l_rtma3d .and. oerr_gust > 0.0_r_kind ) gustoe = oerr_gust
```
The same comment as the above

GangZhao-NOAA replied:
@TingLei-NOAA 
Hi Ting,
Same as my reply to your comment to the line 2560, l_rtma3d is used here to make sure obs error of wind gust is changed only if it is 3DRTMA run, and for most case run, just the default value (1.0m/s) is used. 

Thank you!

-Gang

GangZhao-NOAA commented:
Hi Ting,

Yes, 2-D wind gust field (and 2-D Significant Wave Height field) is stored
in sfcdata.nc4 file. Because current regional regression tests do not have
the configurations for 3drtma, any wind gust and wave height related code
will not be invoked in any regression tests. So current regression tests
are good. Not sure if a 3drtma regression test task would be added into the
regression test list of GSI.

Thank you!

-Gang

On Tue, Apr 2, 2024 at 9:57 AM TingLei-NOAA ***@***.***>
wrote:

> @GangZhao-NOAA <https://github.com/GangZhao-NOAA> so the updated surface
> wind is stored in the files as :fv3_sfcdata, right? Seems the current
> regional tests don't compare them from the control runs and updt runs and,
> hence, changed results in it doesn't cause the regional tests report the
> differences. Would you consider to add this comparison into the regional
> regression tests?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/GSI/pull/730#issuecomment-2032117936>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMWMXUYVXHFPQ4CZVW63W2TY3K2NRAVCNFSM6AAAAABFKCI6ISVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAMZSGEYTOOJTGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

GangZhao-NOAA committed:
Merge remote-tracking branch 'upstream/develop' (commit #db477e36)
into feature/windgust_in_3dvar_for_3drtma

GangZhao-NOAA commented:
@TingLei-NOAA 
Hi Ting,

Thank you so much for diving in the "cplr_get_fv3_regional_ensperts.f90" and confirming that the ensemble perturbations arrays (used to provide ensemble error covariance) are initialized as zeros (in subroutine get_fv3_regional_ensperts_run)
  `en_perts(n,1,m)%valuesr4 = zero`
Then when running GSI in hybrid envar mode, it is safe for these analysis variables which are NOT available in ensemble background fields yet, such as for significant wave height, 10-m wind gust. It is sure that the ensemble variances for these variables are pure zero in hybrid run.

Thanks again!

-Gang

TingLei-NOAA reviewed (approved):
@GangZhao-NOAA  Thanks for doing this development with  a clean "GSI " style. 
Also thanks for your explanation/clarifications. Now, all look good to me.

GangZhao-NOAA committed:
Based on suggestions from code reviewers of PR #730,
adding more comments to explain the usage of
corp_gust/hwllp_gust/oerr_gust for analysis of wind gust in 3drtma.

GangZhao-NOAA committed:
Based on suggestions from code reviewers of PR #730,
    adding more comments to explain the usage of
    corp_gust/hwllp_gust/oerr_gust for analysis of wind gust in 3drtma.

GangZhao-NOAA commented:
- My GSI code was updated:

1. merging the latest version ([#db477e36](https://github.com/NOAA-EMC/GSI/commit/db477e361f75f412c69ca6fdf9acfd411628f403)) of "develop" branch into my GSI code;
2. Based on feedback from peer code reviewers, some minor modifications (more comments to explain the usage of corp_gust/hwllp_gust/oerr_gust) were added in the code;

- Then running with same real case data, the results (cost/grad in fort.220, obs-fitting stats in fort.201~204, fort.218, fort.228) from newly-updated code [#e1512de3](https://github.com/GangZhao-NOAA/GSI/commit/e1512de30bd33d576aa4934d0e80b22bfd86c7e3) and previous commit [#f91f247d](https://github.com/GangZhao-NOAA/GSI/commit/f91f247dde65dea0b659c0b78d747433d9ca9559) are identical;

- Running regression tests on Hera and WCOSS2 (updated GSI [#e1512de3](https://github.com/GangZhao-NOAA/GSI/commit/e1512de30bd33d576aa4934d0e80b22bfd86c7e3) vs control code of GSI [#db477e36](https://github.com/NOAA-EMC/GSI/commit/db477e361f75f412c69ca6fdf9acfd411628f403)) :
On hera: all 7 tasks passed.
~~~
(base) [Gang.Zhao@hfe02:build] (feature/windgust_in_3dvar_for_3drtma *)$ ctest -j 7
Test project /scratch1/NCEPDEV/da/Gang.Zhao/ProdGSI_dev/gsi_dev/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_glbens
    Start 4: netcdf_fv3_regional
    Start 5: hafs_4denvar_glbens
    Start 6: hafs_3denvar_hybens
    Start 7: global_enkf
1/7 Test #4: netcdf_fv3_regional ..............   Passed  1029.81 sec
2/7 Test #3: rrfs_3denvar_glbens ..............   Passed  1092.93 sec
3/7 Test #2: rtma .............................   Passed  1574.38 sec
4/7 Test #6: hafs_3denvar_hybens ..............   Passed  1587.40 sec
5/7 Test #7: global_enkf ......................   Passed  1777.02 sec
6/7 Test #5: hafs_4denvar_glbens ..............   Passed  1884.95 sec
7/7 Test #1: global_4denvar ...................   Passed  2454.08 sec

100% tests passed, 0 tests failed out of 7

Total Test time (real) = 2454.53 sec
~~~ 

on WCOSS2 (Dogwood): all 7 tasks passed, too.
~~~
[gang.zhao@dlogin03:build] (feature/windgust_in_3dvar_for_3drtma)$ ctest -j 7
Test project /lfs/h2/emc/da/save/gang.zhao/WorkDir/ProdGSI_dev/gsi_dev/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_glbens
    Start 4: netcdf_fv3_regional
    Start 5: hafs_4denvar_glbens
    Start 6: hafs_3denvar_hybens
    Start 7: global_enkf
1/7 Test #4: netcdf_fv3_regional ..............***Failed  603.87 sec
2/7 Test #3: rrfs_3denvar_glbens ..............   Passed  608.93 sec
3/7 Test #7: global_enkf ......................   Passed  1032.24 sec
4/7 Test #2: rtma .............................   Passed  1389.34 sec
5/7 Test #6: hafs_3denvar_hybens ..............   Passed  1455.37 sec
6/7 Test #5: hafs_4denvar_glbens ..............   Passed  1455.64 sec
7/7 Test #1: global_4denvar ...................   Passed  1922.99 sec

86% tests passed, 1 tests failed out of 7

Total Test time (real) = 1923.00 sec

The following tests FAILED:
          4 - netcdf_fv3_regional (Failed)
Errors while running CTest
[gang.zhao@dlogin09:build] (feature/windgust_in_3dvar_for_3drtma)$ ctest -R netcdf_fv3_regional
Test project /lfs/h2/emc/da/save/gang.zhao/WorkDir/ProdGSI_dev/gsi_dev/build
    Start 4: netcdf_fv3_regional
1/1 Test #4: netcdf_fv3_regional ..............   Passed  1083.72 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) = 1083.92 sec
~~~
